spec/lang/representation.md

```diff
@@ -131,25 +131,28 @@ fn encode_ptr<M: Memory>(ptr: ThinPointer<M::Provenance>) -> List<AbstractByte<M
 }
 
 impl PointerMetaKind {
-    /// Returns the representable type of the metadata if there is any.
-    pub fn ty<T: Target>(self) -> Option<Type> {
+    /// Returns the representable type of the metadata.
+    pub fn ty<T: Target>(self) -> Type {
         match self {
-            PointerMetaKind::None => None,
-            PointerMetaKind::ElementCount => Some(Type::Int(IntType::usize_ty::<T>())),
-            PointerMetaKind::VTablePointer(..) => Some(Type::Ptr(PtrType::VTablePtr)),
+            PointerMetaKind::None => unit_type(),
+            PointerMetaKind::ElementCount => Type::Int(IntType::usize_ty::<T>()),
+            PointerMetaKind::VTablePointer(..) => Type::Ptr(PtrType::VTablePtr),
         }
     }
 }
 
 impl PtrType {
     /// Returns a pair type representing this wide pointer or `None` if it is thin.
     pub fn as_wide_pair<T: Target>(self) -> Option<Type> {
-        let meta_ty = self.meta_kind().ty::<T>()?;
+        if self.meta_kind() == PointerMetaKind::None {
+            return None;
+        }
```

Comment:
Hm, this is not pretty... where is this function used?

Reply:
How so? Previously it was that `ty()` returned None for thin pointers, now the as_wide_pair does this, which is used to encode wide pointers as tuples (but not thin pointers, as this would lead to infinite recursion).

Reply:
Previously we didn't need an `if` like this, did we? The `?` handled this implicitly which is nicer.

Given all the `.unwrap_or_else(unit_type)` I think the new way is still better. I also can't find a nicer way to organize things in or around `as_wide_pair` here so let's just keep this.